x-pack/lib/config_management/elasticsearch_source.rb

```diff
@@ -193,5 +191,54 @@ def client
         @client ||= build_client
       end
     end
+
+    class SystemIndicesFetcher
+      include LogStash::Util::Loggable
+
+      SYSTEM_INDICES_API_PATH = "_logstash/pipeline"
+
+      def fetch_config(pipeline_ids, client)
+        path_ids = pipeline_ids.join(",")
+        client.get("#{SYSTEM_INDICES_API_PATH}/#{path_ids}")
+      end
+
+      def format_response(response)
+        # for backwards compatibility
+        response
+      end
+
+      def get_single_pipeline_setting(response, pipeline_id)
+        response.fetch(pipeline_id, {})
+      end
+    end
+
+    # TODO clean up LegacyHiddenIndicesFetcher when 7.9.* is deprecated
+    class LegacyHiddenIndicesFetcher
+      include LogStash::Util::Loggable
+
+      PIPELINE_INDEX = ".logstash"
+
+      def fetch_config(pipeline_ids, client)
+        request_body_string = LogStash::Json.dump({ "docs" => pipeline_ids.collect { |pipeline_id| { "_id" => pipeline_id } } })
+        client.post("#{PIPELINE_INDEX}/_mget", {}, request_body_string)
+      end
+
+      def format_response(response)
+        if response["docs"].nil?
+          logger.debug("Server returned an unknown or malformed document structure", :response => response)
+          raise ElasticsearchSource::RemoteConfigError, "Elasticsearch returned an unknown or malformed document structure"
+        end
+
+        response["docs"].map { |pipeline|
+          {pipeline["_id"] => pipeline} if pipeline.fetch("found", false)
+        }.compact
+        .reduce({}, :merge)
```

Comment:
The legacy response has a different json structure. To make both APIs following the same flow, I format the legacy json to look like the new one. 
`{ "docs": [{}, {}] }`  to  `{ "pipeline1": {}, "pipeline2": {} }`